File: srat_henz.py
"""Self-Referential Aptitude Test from Henz 1996, used as a worked example
in the Mozart/Oz documentation (10Q × 5).
https://mozart.github.io/mozart-v1/doc-1.4.0/fdt/node38.html
"""

from solver import report

A, B, C, D, E = 1, 2, 3, 4, 5
PRIMES = {2, 3, 5, 7}
SQUARES = {0, 1, 4, 9}
CUBES = {0, 1, 8}
FACTORIALS = {1, 2, 6}
VALS = [A, B, C, D, E]


def run(amb, fail):
    def at_6(q1, q2, q3, q4, q5, q6):
        q = [None, q1, q2, q3, q4, q5, q6]
        # Q1: first B at q[1] + 1.
        t1 = q1 + 1
        if q[t1] != B: return fail()
        if B in q[1:t1]: return fail()
        return amb(VALS, lambda q7: at_7(q1, q2, q3, q4, q5, q6, q7))

    def at_7(q1, q2, q3, q4, q5, q6, q7):
        q = [None, q1, q2, q3, q4, q5, q6, q7]
        # Q3: q[3] = q[t3]; t3 = [1,2,4,7,6][q[3]-1].
        t3 = [1, 2, 4, 7, 6][q3 - 1]
        if q[t3] != q3: return fail()
        return amb(VALS, lambda q8: at_8(q1, q2, q3, q4, q5, q6, q7, q8))

    def at_8(q1, q2, q3, q4, q5, q6, q7, q8):
        # Q7: |q[7]-q[8]| = 5 - q[7].
        if abs(q7 - q8) != 5 - q7: return fail()
        return amb(VALS, lambda q9:
            amb(VALS, lambda q10:
                at_10(q1, q2, q3, q4, q5, q6, q7, q8, q9, q10)))

    def at_10(q1, q2, q3, q4, q5, q6, q7, q8, q9, q10):
        q = [None, q1, q2, q3, q4, q5, q6, q7, q8, q9, q10]
        count = lambda pred: sum(1 for v in q[1:] if pred(v))
        # Q2: only consecutive equal pair at pairs[q[2]-1].
        pa, pb = [(2,3),(3,4),(4,5),(5,6),(6,7)][q2 - 1]
        if q[pa] != q[pb]: return fail()
        if any(q[i] == q[i + 1] for i in range(1, 10) if i != pa):
            return fail()
        # Q4: count(A) = q[4] - 1.
        ca = count(lambda v: v == A)
        if ca != q4 - 1: return fail()
        # Q5: q[5] = q[11 - q[5]].
        if q5 != q[11 - q5]: return fail()
        # Q6: count(A) = count(letter[q[6]+1]); E means none match.
        if q6 == E:
            if any(count(lambda x, v=v: x == v) == ca for v in (B,C,D,E)):
                return fail()
        elif count(lambda v, t=q6 + 1: v == t) != ca:
            return fail()
        # Q8: count(vowels) = q[8] + 1.
        if count(lambda v: v == A or v == E) != q8 + 1: return fail()
        # Q9: consonant count is prime/factorial/square/cube/÷5.
        cc = count(lambda v: v in (B, C, D))
        ok9 = (cc in PRIMES if q9 == A
               else cc in FACTORIALS if q9 == B
               else cc in SQUARES if q9 == C
               else cc in CUBES if q9 == D
               else cc % 5 == 0)
        if not ok9: return fail()
        # Q10: tautology.
        return q

    return amb(VALS, lambda q1:
        amb(VALS, lambda q2:
        amb(VALS, lambda q3:
        amb(VALS, lambda q4:
        amb(VALS, lambda q5:
        amb(VALS, lambda q6: at_6(q1, q2, q3, q4, q5, q6)))))))


puzzle = {"name": "SRAT (Henz, 10Q × 5)", "run": run}

if __name__ == "__main__":
    report(puzzle)

File: srat-henz.js
// Self-Referential Aptitude Test from Henz 1996, used as a worked example
// in the Mozart/Oz documentation (10Q × 5).
// https://mozart.github.io/mozart-v1/doc-1.4.0/fdt/node38.html

import { report } from "./solver";

const A = 1, B = 2, C = 3, D = 4, E = 5;
const PRIMES = new Set([2, 3, 5, 7]);
const SQUARES = new Set([0, 1, 4, 9]);
const CUBES = new Set([0, 1, 8]);
const FACTORIALS = new Set([1, 2, 6]);
const VALS = [A, B, C, D, E];

const puzzle = {
  name: "SRAT (Henz, 10Q × 5)",
  run: (amb, fail) =>
    amb(VALS, q1 =>
    amb(VALS, q2 =>
    amb(VALS, q3 =>
    amb(VALS, q4 =>
    amb(VALS, q5 =>
    amb(VALS, q6 => {
      const q16 = [null, q1, q2, q3, q4, q5, q6];
      // Q1: first B at q[1] + 1.
      const t1 = q1 + 1;
      if (q16[t1] !== B) return fail();
      if (q16.slice(1, t1).includes(B)) return fail();
      return amb(VALS, q7 => {
        // Q3: q[3] = q[t3]; t3 = [1,2,4,7,6][q[3]-1].
        const q17 = [null, q1, q2, q3, q4, q5, q6, q7];
        const t3 = [1, 2, 4, 7, 6][q3 - 1];
        if (q17[t3] !== q3) return fail();
        return amb(VALS, q8 => {
          // Q7: |q[7]-q[8]| = 5 - q[7].
          if (Math.abs(q7 - q8) !== 5 - q7) return fail();
          return amb(VALS, q9 =>
          amb(VALS, q10 => {
            const q = [null, q1, q2, q3, q4, q5, q6, q7, q8, q9, q10];
            const count = pred => q.slice(1).filter(pred).length;

            // Q2: only consecutive equal pair at pairs[q[2]-1].
            const [pa, pb] = [[2,3],[3,4],[4,5],[5,6],[6,7]][q2 - 1];
            if (q[pa] !== q[pb]) return fail();
            for (let i = 1; i < 10; i++) {
              if (i !== pa && q[i] === q[i + 1]) return fail();
            }
            // Q4: count(A) = q[4] - 1.
            const ca = count(v => v === A);
            if (ca !== q4 - 1) return fail();
            // Q5: q[5] = q[11 - q[5]].
            if (q5 !== q[11 - q5]) return fail();
            // Q6: count(A) = count(letter[q[6]+1]); E means none match.
            if (q6 === E) {
              if ([B,C,D,E].some(v => count(x => x === v) === ca))
                return fail();
            } else if (count(v => v === q6 + 1) !== ca) return fail();
            // Q8: count(vowels) = q[8] + 1.
            if (count(v => v === A || v === E) !== q8 + 1) return fail();
            // Q9: consonant count is prime/factorial/square/cube/÷5.
            const cc = count(v => v === B || v === C || v === D);
            const ok9 = q9 === A ? PRIMES.has(cc)
                      : q9 === B ? FACTORIALS.has(cc)
                      : q9 === C ? SQUARES.has(cc)
                      : q9 === D ? CUBES.has(cc)
                      :            cc % 5 === 0;
            if (!ok9) return fail();
            // Q10: tautology.
            return q;
          }));
        });
      });
    })))))),
};

export default puzzle;
import.meta.main && report(puzzle);
